packages/lit-html/src/lib/lit-html.ts

```diff
@@ -12,9 +12,548 @@
+ */
+class TemplateInstance {
+  __template: Template;
+  __parts: Array<Part | undefined> = [];
+
+  constructor(template: Template) {
+    this.__template = template;
+  }
+
+  // This method is separate from the constructor because we need to return a
+  // DocumentFragment and we don't want to hold onto it with an instance field.
+  __clone(options: RenderOptions | undefined) {
+    const {
+      __element: {content},
+      __parts: parts,
+    } = this.__template;
+    const fragment = d.importNode(content, true);
+    walker.currentNode = fragment;
+
+    let node = walker.nextNode();
+    let nodeIndex = 0;
+    let partIndex = 0;
+    let templatePart = parts[0];
+
+    while (templatePart !== undefined && node !== null) {
+      if (nodeIndex === templatePart.index) {
+        let part: Part | undefined;
+        if (templatePart.type === NODE_PART) {
+          part = new NodePart(node as HTMLElement, node.nextSibling, options);
+        } else {
+          // TODO: other part types
+        }
+        this.__parts.push(part);
+        templatePart = parts[++partIndex];
+      }
+      if (templatePart !== undefined && nodeIndex !== templatePart.index) {
+        node = walker.nextNode();
+        nodeIndex++;
+      }
+    }
+    return fragment;
+  }
+
+  __update(values: Array<unknown>) {
+    let i = 0;
+    for (const part of this.__parts) {
+      if (part === undefined) {
+        i++;
+        continue;
+      }
+      // TODO: other part types
+      (part as NodePart).__setValue(values[i++]);
+    }
+  }
+}
+
+/*
+ * Parts
+ */
+
+const ATTRIBUTE_PART = 1;
+const NODE_PART = 2;
+const ELEMENT_PART = 3;
+const COMMENT_PART = 4;
+
+type AttributeTemplatePart = {
+  readonly type: typeof ATTRIBUTE_PART;
+  index: number;
+  readonly name: string;
+  readonly strings: ReadonlyArray<string>;
+};
+type NodeTemplatePart = {
+  readonly type: typeof NODE_PART;
+  index: number;
+};
+type ElementTemplatePart = {
+  readonly type: typeof ELEMENT_PART;
+  index: number;
+};
+type CommentTemplatePart = {
+  readonly type: typeof COMMENT_PART;
+  index: number;
+};
+
+/**
+ * A TemplatePart represents a dynamic part in a template, before the template
+ * is instantiated. When a template is instantiated Parts are created from
+ * TemplateParts.
+ */
+type TemplatePart =
+  | NodeTemplatePart
+  | AttributeTemplatePart
+  | ElementTemplatePart
+  | CommentTemplatePart;
+
+export type Part = NodePart;
+
+export class NodePart {
+  __value: unknown;
+  constructor(
+    public __startNode: ChildNode,
+    public __endNode: ChildNode | null,
+    public options: RenderOptions | undefined
+  ) {}
+
+  __setValue(value: unknown): void {
+    if (isPrimitive(value)) {
+      if (value !== this.__value) {
+        this.__commitText(value);
+      }
+    } else if ((value as TemplateResult)._$litType$ !== undefined) {
+      this.__commitTemplateResult(value as TemplateResult);
+    } else if ((value as Node).nodeType !== undefined) {
```

Comment:
Yeah, this is for perf. Jason Miller had a esbench somewhere that showed the fastest ways to detect a node, and this was one of them. I agree, this could be a false positive. Something to look out for?